File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/rsp/RspScan.kt
package com.mytiki.sdk.capture.receipt.capacitor.rsp

import com.getcapacitor.JSObject
import com.microblink.core.ScanResults
import com.mytiki.sdk.capture.receipt.capacitor.Account
import org.json.JSONObject

/**
 * Represents a result of the RSP (Receipt Scanning Process) scan operation.
 *
 * @property _account The associated [Account] information if available, or null.
 * @property _scan The scanned receipt information if available, or null.
 * @property _isRunning Indicates whether the scanning process is still running.
 */
class RspScan(
    requestId: String,
    scan: ScanResults?,
    account: Account? = null,
    isRunning: Boolean = true
) : Rsp(requestId) {
    private val _account: Account? = account
    private val _scan: RspReceipt? = RspReceipt.opt(requestId, scan)
    private val _isRunning: Boolean = isRunning

    /**
     * Converts this [RspScan] object to a JSON representation.
     *
     * @return A [JSONObject] containing the JSON representation of this [RspScan] object.
     */
    override fun toJS(): JSObject =
        JSObject()
            .put("account", if (_account != null) RspAccount(this.requestId, _account).toJS() else null)
            .put("receipt", _scan?.toJS())
            .put("isRunning", _isRunning)
}

File: ios/Plugin/Plugin/Rsp/RspScan.swift
/*
 * RspScan Struct
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

import Foundation
import Capacitor

/**
 Represents a response containing scan information.

 This struct is used to convey details about a scan, including the scanned receipt, whether the scan is running, and additional account information if available.
 */
public struct RspScan{
    var requestId: String 
    var event: PluginEvent
    /// The scanned receipt information.
    let scan: RspReceipt
    /// Indicates whether the scan is currently running.
    let isRunning: Bool = false
    /// Additional account information, if available.
    var account: Account? = nil
    
    /**
     Initializes an `RspScan` struct.

     - Parameters:
        - scan: The scanned receipt information.
        - account: Additional account information (optional, default is `nil`).
     */
    public init(requestId: String, event: PluginEvent, scan: RspReceipt, account: Account? = nil) {
        self.scan = scan
        self.account = account
        self.requestId = requestId
        self.event = event
    }
    /**
     Converts the `RspScan` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing scan information in a format suitable for a Capacitor plugin call result.
     */
    func toPluginCallResultData() -> Capacitor.PluginCallResultData {
        return [
            "account" : account?.toResultData() as Any,
            "scan" : scan.toPluginCallResultData(),
            "isRunning" : isRunning
        ]
            
    }
    
}
